crates/ra_hir/src/ty/infer.rs

```diff
@@ -483,9 +497,40 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             Resolution::Def(def) => {
                 let typable: Option<TypableDef> = def.into();
                 let typable = typable?;
+
+                let ty = self.db.type_for_def(typable, Namespace::Values);
                 let substs = Ty::substs_from_path(self.db, &self.resolver, path, typable);
-                let ty = self.db.type_for_def(typable, Namespace::Values).subst(&substs);
+                let ty = ty.subst(&substs);
                 let ty = self.insert_type_vars(ty);
+
+                // plug the old parent_ty in
+                if let Some(actual_def_ty) = actual_def_ty {
+                    if let crate::ModuleDef::Function(func) = def {
+                        let gen = func.generic_params(self.db);
+                        if let Some(target_ty) = func.impl_block(self.db) {
+                            let target_ty = target_ty.target_ty(self.db);
+                            let old_params = target_ty.substs().unwrap().clone();
+
+                            let target_ty = target_ty.subst(&substs);
+                            let target_ty = self.insert_type_vars(target_ty);
+
+                            if gen.count_parent_params() > 0 {
+                                self.unify(&target_ty, &actual_def_ty);
+
+                                if let Ty::Apply(ty) = &ty {
```

Comment:
I'm not sure what the purpose of this block is. Isn't the unification above enough?

Reply:
The code above unify `Gen::<u32,u64>" and `Gen::<U,V>`, and the block is unifying following function  :
` ``
impl<T> Gen<u32,T> {
  fn make<T>();            <- this function
}
``` 
We have to "link up" the actual function and its impl block target.